internal/common/resource.go

```diff
@@ -36,24 +36,87 @@ func CollectResourceStatus(request *Request, funcs ...ReconcileFunc) ([]Resource
 type ResourceUpdateFunc = func(expected, found controllerutil.Object)
 type ResourceStatusFunc = func(resource controllerutil.Object) ResourceStatus
 
-func CreateOrUpdateResource(request *Request, resource controllerutil.Object, updateResource ResourceUpdateFunc) (ResourceStatus, error) {
-	return createOrUpdate(request, resource, false, updateResource, statusOk)
+type ReconcileBuilder interface {
+	NamespacedResource(controllerutil.Object) ReconcileBuilder
+	ClusterResource(controllerutil.Object) ReconcileBuilder
+	WithAppLabels(name string, component AppComponent) ReconcileBuilder
+	UpdateFunc(ResourceUpdateFunc) ReconcileBuilder
+	StatusFunc(ResourceStatusFunc) ReconcileBuilder
+
+	Reconcile() (ResourceStatus, error)
 }
 
-func CreateOrUpdateResourceWithStatus(request *Request, resource controllerutil.Object, updateResource ResourceUpdateFunc, statusFunc ResourceStatusFunc) (ResourceStatus, error) {
-	return createOrUpdate(request, resource, false, updateResource, statusFunc)
+type reconcileBuilder struct {
+	request           *Request
+	resource          controllerutil.Object
+	isClusterResource bool
+
+	addLabels        bool
+	operandName      string
+	operandComponent AppComponent
+
+	updateFunc ResourceUpdateFunc
+	statusFunc ResourceStatusFunc
+}
+
+var _ ReconcileBuilder = &reconcileBuilder{}
+
+func (r *reconcileBuilder) NamespacedResource(resource controllerutil.Object) ReconcileBuilder {
```

Comment:
I would split that into:
`Resource(resource controllerutil.Object)...`
and
`Namespaced()...` or `Namespaced(bool)` instead.

Reply:
I did not split it, because this way it is not possible to forget to specify if a resource is cluster or namespaced.
I think in a future patch we may remove the distinction between cluster and namespaced resources and use owner annotations for all of them to simplify the code..